programs/compressor/Compressor.cpp

```diff
@@ -132,25 +141,52 @@ int mainEntryClickHouseCompressor(int argc, char ** argv)
             codec = CompressionCodecFactory::instance().get(method_family, level);
 
 
-        ReadBufferFromFileDescriptor rb(STDIN_FILENO);
-        WriteBufferFromFileDescriptor wb(STDOUT_FILENO);
+        std::unique_ptr<ReadBufferFromFileBase> rb;
+        std::unique_ptr<WriteBufferFromFileBase> wb;
+
+        if (options.count("input"))
+            rb = std::make_unique<ReadBufferFromFile>(options["input"].as<std::string>());
+        else
+            rb = std::make_unique<ReadBufferFromFileDescriptor>(STDIN_FILENO);
+
+        if (options.count("output"))
+            wb = std::make_unique<WriteBufferFromFile>(options["output"].as<std::string>());
+        else
+            wb = std::make_unique<WriteBufferFromFileDescriptor>(STDOUT_FILENO);
 
         if (stat_mode)
         {
             /// Output statistic for compressed file.
-            checkAndWriteHeader(rb, wb);
+            checkAndWriteHeader(*rb, *wb);
         }
         else if (decompress)
         {
             /// Decompression
-            CompressedReadBuffer from(rb);
-            copyData(from, wb);
+
+            size_t offset_in_compressed_file = options["offset-in-compressed-file"].as<size_t>();
+            size_t offset_in_decompressed_block = options["offset-in-decompressed-block"].as<size_t>();
+
+            if (offset_in_compressed_file || offset_in_decompressed_block)
+            {
+                if (!options.count("input"))
+                {
+                    throw DB::Exception("--offset-in-compressed-file/--offset-in-decompressed-block requires --input", DB::ErrorCodes::BAD_ARGUMENTS);
```

Comment:
It's not necessarily.

Reply:
Maybe it's required by the interface of CompressedReadBuffer/CompressedReadBufferFromFile?

Reply:
Not really, and indeed this is not 100% guarantee that this will work, for example pipes via process substitution will not work obviously, i.e. 

```
$ clickhouse-compressor --decompress --offset-in-compressed-file 1 <(clickhouse-compressor <<<foo)
Code: 87, e.displayText() = DB::ErrnoException: Cannot seek through file /dev/fd/63, errno: 29, strerror: Illegal seek, Stack trace (when copying this message, always include the lines below):
```

So it is better to remove this doubtful condition (I don't think that it worth to check for `S_IFREG`, failed `seek` looks good enough for this utility)

Reply:
Yes, just try to do seek without any checks - if it will throw exception, it is explanatory enough.

Reply:
https://github.com/ClickHouse/ClickHouse/pull/18675